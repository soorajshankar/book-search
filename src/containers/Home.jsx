import React, {useState} from 'react'
import Button from '../components/Button'
import Container from '../components/Container'
import BookList from '../components/BookList'
import AutoComplete from '../components/AutoComplete'
import {getSuggestions} from '../utils/UiUtil'
import {data} from '../utils/data'
import {searchSummary} from '../utils/utils'

const Home = () => {
  const [value, setValue] = useState('')
  const [maxSug, setMaxSug] = useState(5) // Maximum suggestions displayed
  const [suggestions, setSuggestions] = useState([])
  const [results, setResults] = useState([])

  const onChange = e => {
    const input = e.target.value
    if (input.length < 1) return setSuggestions([])
    const suggestions = getSuggestions(e.target.value, data.queries)
    setSuggestions([...suggestions])
    setValue(input)
  }

  const onSubmit = () => {
    const result = searchSummary(
      value,
      data.summaries,
      maxSug,
      data.titles,
      data.authors,
    )
    setResults(result)
  }
  const onSelect = e => {
    setSuggestions([])
    setValue(e)
  }
  const handleMaxSug = e => setMaxSug(e.target.value)
  return (
    <Container>
      <h2 className="c-title">Search Books</h2>
      <div className="c-container__shell">
        <AutoComplete
          {...{onChange, onSelect, visible: true, suggestions, value}}
        />
        <Button className="btn__search" onClick={onSubmit}>
          Search
        </Button>
      </div>
      <MaxSuggestions {...{maxSug, handleMaxSug}} />
      <BookList list={results} />
    </Container>
  )
}
const MaxSuggestions = ({maxSug, handleMaxSug}) => (
  <div className="c-drawer__item">
    <label>Max. suggestions : </label>
    <input
      value={maxSug}
      onChange={handleMaxSug}
      style={{width: 50}}
      type="number"
    />
  </div>
)

export default Home
