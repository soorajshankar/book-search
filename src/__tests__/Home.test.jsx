import React from 'react'
import {render, fireEvent} from '@testing-library/react'
import Home from '../containers/Home'
import {act} from 'react-dom/test-utils'

describe('renders Home', () => {
  it('should match Snapshot', () => {
    let count = 0
    const {container} = render(<Home />)
    expect(container).toMatchSnapshot()
  })
  it('should render with no value', () => {
    let count = 0
    const testMessage = 'No books Selected'
    const {container, getByText} = render(<Home />)
    expect(container.firstChild.className).toEqual('App')

    // input
    const inputDom = container.querySelector('.c-ac__input')
    expect(inputDom.value).toEqual('')

    // suggestions
    const listDom = container.querySelector('.c-ac__suggestions')
    expect(listDom).toBeNull()

    expect(getByText(testMessage)).toBeInTheDocument()
  })
  it('should render values onChange', async () => {
    let count = 0
    const {container} = render(<Home />)

    expect(container.firstChild.className).toEqual('App')

    // input
    const inputDom = container.querySelector('.c-ac__input')
    expect(inputDom.value).toEqual('')

    // suggestions
    const listDom = container.querySelector('.c-ac__suggestions')
    expect(listDom).toBeNull()

    // input event
    fireEvent.change(inputDom, {
      target: {value: 'is'},
    })
    await act(async () => await sleep(1000)) // debounce time
    expect(inputDom.value).toEqual('is')
  })
})

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms))
}
