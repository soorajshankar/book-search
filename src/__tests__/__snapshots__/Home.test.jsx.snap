// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Home should match Snapshot 1`] = `
<div>
  <div
    class="App"
    style="margin: 20px;"
  >
    <h2
      class="c-title"
    >
      Search Books
    </h2>
    <div
      class="c-container__shell"
    >
      <div
        class="c-ac"
      >
        <input
          class="c-ac__input"
          value=""
        />
      </div>
      <button
        class="btn__search"
      >
        Submit
      </button>
    </div>
    <div
      class="c-drawer__item"
    >
      <label>
        Max. suggestions : 
      </label>
      <input
        style="width: 50px;"
        type="number"
        value="5"
      />
    </div>
    <div
      class="c-list"
      data-testid="book_list"
    >
      <h4
        style="color: rgb(136, 136, 132);"
      >
        No books Selected
      </h4>
    </div>
  </div>
</div>
`;
